File: ga/glm4_result_2024_6/result/tree_glm/40_tree.c
use std::rc::Rc;
use std::rc::Weak;
use std::cell::RefCell;

type TreeLink = Option<Rc<RefCell<TreeNode>>>;

#[derive(Debug)]
struct TreeNode {
    element: i32,
    left: TreeLink,
    right: TreeLink,
}

impl TreeNode {
    fn new(element: i32) -> Rc<RefCell<TreeNode>> {
        Rc::new(RefCell::new(TreeNode {
            element,
            left: None,
            right: None,
        }))
    }
}

fn make_empty(tree: &mut TreeLink) {
    if let Some(node) = tree.take() {
        make_empty(&mut node.borrow_mut().left);
        make_empty(&mut node.borrow_mut().right);
    }
}

fn find(x: i32, tree: &TreeLink) -> TreeLink {
    let mut current = tree.clone();
    while let Some(n) = current {
        if x < n.borrow().element {
            current = n.borrow_mut().left.clone();
        } else if x > n.borrow().element {
            current = n.borrow_mut().right.clone();
        } else {
            return Some(n);
        }
    }
    None
}

fn find_min(tree: &TreeLink) -> TreeLink {
    let mut current = tree.clone();
    while let Some(n) = current {
        if n.borrow().left.is_none() {
            return Some(n);
        }
        current = n.borrow_mut().left.clone();
    }
    None
}

fn find_max(tree: &TreeLink) -> TreeLink {
    let mut current = tree.clone();
    while let Some(n) = current {
        if n.borrow().right.is_none() {
            return Some(n);
        }
        current = n.borrow_mut().right.clone();
    }
    None
}

fn insert(x: i32, tree: &mut TreeLink) {
    if let Some(n) = tree.clone() {
        if x < n.borrow().element {
            insert(x, &mut n.borrow_mut().left);
        } else if x > n.borrow().element {
            insert(x, &mut n.borrow_mut().right);
        }
    } else {
        *tree = Some(TreeNode::new(x));
    }
}

fn delete(x: i32, tree: &mut TreeLink) {
    if let Some(mut n) = tree.take() {
        if x < n.borrow().element {
            delete(x, &mut n.borrow_mut().left);
            *tree = Some(n);
        } else if x > n.borrow().element {
            delete(x, &mut n.borrow_mut().right);
            *tree = Some(n);
        } else {
            if let Some(mut right) = n.borrow_mut().right.take() {
                if let Some(min) = find_min(&right) {
                    let mut min = min.borrow_mut();
                    if let Some(mut left) = min.left.take() {
                        min.left = Some(left);
                    }
                    min.right = n.borrow_mut().left.take();
                    *tree = Some(min);
                } else {
                    *tree = Some(right);
                }
            } else {
                *tree = n.borrow_mut().left.take();
            }
        }
    }
}

fn retrieve(p: &TreeLink) -> i32 {
    p.as_ref().unwrap().borrow().element
}

fn main() {
    let mut tree: TreeLink = None;
    for i in 0..50 {
        insert(i, &mut tree);
    }
    for i in 0..50 {
        let p = find(i, &tree);
        if p.is_none() || retrieve(&p.unwrap()) != i {
            println!("Error at {}", i);
        }
    }
    for i in 0..50 {
        if i % 2 == 0 {
            delete(i, &mut tree);
        }
    }
    for i in 1..50 {
        if i % 2 == 1 {
            let p = find(i, &tree);
            if p.is_none() || retrieve(&p.unwrap()) != i {
                println!("Error at {}", i);
            }
        } else {
            let p = find(i, &tree);
            if p.is_some() {
                println!("Error at {}", i);
            }
        }
    }
    println!("Min is {}, Max is {}", retrieve(&find_min(&tree).unwrap()), retrieve(&find_max(&tree).unwrap()));
}

File: ga/4omini_result_2024_7/result/tree_mini/40_tree.rs
use std::cmp::Ord;
use std::collections::HashSet;

#[derive(Debug)]
struct TreeNode {
    element: i32,
    left: Option<Box<TreeNode>>,
    right: Option<Box<TreeNode>>,
}

struct SearchTree {
    root: Option<Box<TreeNode>>,
}

impl SearchTree {
    fn new() -> Self {
        SearchTree { root: None }
    }

    fn make_empty(&mut self) {
        self.root = None;
    }

    fn find(&self, x: i32) -> Option<&TreeNode> {
        Self::find_node(&self.root, x)
    }

    fn find_node(node: &Option<Box<TreeNode>>, x: i32) -> Option<&TreeNode> {
        match node {
            Some(n) => {
                if x < n.element {
                    Self::find_node(&n.left, x)
                } else if x > n.element {
                    Self::find_node(&n.right, x)
                } else {
                    Some(n)
                }
            }
            None => None,
        }
    }

    fn find_min(&self) -> Option<&TreeNode> {
        Self::find_min_node(&self.root)
    }

    fn find_min_node(node: &Option<Box<TreeNode>>) -> Option<&TreeNode> {
        match node {
            Some(n) => {
                if n.left.is_none() {
                    Some(n)
                } else {
                    Self::find_min_node(&n.left)
                }
            }
            None => None,
        }
    }

    fn find_max(&self) -> Option<&TreeNode> {
        Self::find_max_node(&self.root)
    }

    fn find_max_node(node: &Option<Box<TreeNode>>) -> Option<&TreeNode> {
        match node {
            Some(n) => {
                if n.right.is_none() {
                    Some(n)
                } else {
                    Self::find_max_node(&n.right)
                }
            }
            None => None,
        }
    }

    fn insert(&mut self, x: i32) {
        self.root = Self::insert_node(self.root.take(), x);
    }

    fn insert_node(node: Option<Box<TreeNode>>, x: i32) -> Option<Box<TreeNode>> {
        match node {
            None => {
                Some(Box::new(TreeNode {
                    element: x,
                    left: None,
                    right: None,
                }))
            }
            Some(mut n) => {
                if x < n.element {
                    n.left = Self::insert_node(n.left.take(), x);
                } else if x > n.element {
                    n.right = Self::insert_node(n.right.take(), x);
                }
                Some(n)
            }
        }
    }

    fn delete(&mut self, x: i32) {
        self.root = Self::delete_node(self.root.take(), x);
    }

    fn delete_node(node: Option<Box<TreeNode>>, x: i32) -> Option<Box<TreeNode>> {
        match node {
            None => {
                eprintln!("Error: Element not found");
                return None;
            }
            Some(mut n) => {
                if x < n.element {
                    n.left = Self::delete_node(n.left.take(), x);
                } else if x > n.element {
                    n.right = Self::delete_node(n.right.take(), x);
                } else {
                    if n.left.is_some() && n.right.is_some() {
                        let min_node = Self::find_min_node(&n.right).unwrap();
                        n.element = min_node.element;
                        n.right = Self::delete_node(n.right.take(), min_node.element);
                    } else {
                        return if n.left.is_none() {
                            n.right.take()
                        } else {
                            n.left.take()
                        };
                    }
                }
                Some(n)
            }
        }
    }
}

fn main() {
    let mut tree = SearchTree::new();
    let mut j = 0;

    for _ in 0..50 {
        tree.insert(j);
        j = (j + 7) % 50;
    }

    // Verify insertion
    for i in 0..50 {
        if tree.find(i).is_none() {
            println!("Error at {}", i);
        }
    }

    // Delete even numbers
    for i in (0..50).step_by(2) {
        tree.delete(i);
    }

    // Verify deletion
    for i in (1..50).step_by(2) {
        if tree.find(i).is_none() {
            println!("Error at {}", i);
        }
    }

    for i in (0..50).step_by(2) {
        if tree.find(i).is_some() {
            println!("Error at {}", i);
        }
    }

    // Print min and max
    let min_value = tree.find_min().map_or("None".to_string(), |n| n.element.to_string());
    let max_value = tree.find_max().map_or("None".to_string(), |n| n.element.to_string());
    println!("Min is {}, Max is {}", min_value, max_value);
}